### ISSUE TITLE
Extracting Smiles for fingerprint bits in featurization

### ISSUE THREAD

**Issue by tobiasharren** [deepchem/deepchem#1761]
- DeepChem Version: 2.3
- Platform: Linux

I ran across some issues discussed in this issues, regarding the smiles environments of size 1 and was wondering, if it should be implemented in deepchem
https://github.com/deepchem/deepchem/issues/897

I commented there, but I think it probably is going to get lost, since the issue is already closed.
The problem is that environments with radius 0 don't create smiles, but they might be relevant for interpretation purposes nonetheless. My propose solution would be:

>   def _featurize(self, mol):
    """
    Calculate circular fingerprint.

    Parameters
    ----------
    mol : RDKit Mol
        Molecule.
    """
    from rdkit import Chem
    from rdkit.Chem import rdMolDescriptors
    if self.sparse:
      info = {}
      fp = rdMolDescriptors.GetMorganFingerprint(
          mol,
          self.radius,
          useChirality=self.chiral,
          useBondTypes=self.bonds,
          useFeatures=self.features,
          bitInfo=info)
      fp = fp.GetNonzeroElements()  # convert to a dict

      # generate SMILES for fragments
      if self.smiles:
        fp_smiles = {}
        for fragment_id, count in fp.items():
          root, radius = info[fragment_id][0]
          if radius == 0:
            atom = mol.GetAtomWithIdx(root)
            aromatic = atom.GetIsAromatic()
            symbol = atom.GetSymbol()
            if aromatic:
                symbol = symbol.lower()
            smiles = "[" + symbol + "]"
          else:
            env = Chem.FindAtomEnvironmentOfRadiusN(mol, radius, root)
            frag = Chem.PathToSubmol(mol, env)
            smiles = Chem.MolToSmiles(frag)
          fp_smiles[fragment_id] = {'smiles': smiles, 'count': count}
        fp = fp_smiles
    else:
      fp = rdMolDescriptors.GetMorganFingerprintAsBitVect(
          mol,
          self.radius,
          nBits=self.size,
          useChirality=self.chiral,
          useBondTypes=self.bonds,
          useFeatures=self.features)
    return fp

In addition I would be interested in extracting fragments associated with bits in the neural fingerprint layer of graph conv networks. (In the Duvenaud Neural Fingerprint paper they extracted the features, but I wasn't able to identify how that worked) Do you have an idea how that could be done?

best regards
Tobias

**Comment by rbharath** [comment]
What would an environment of radius 0 correspond to? Would that just be an atom? If there's a clear use case for this, I'd be pretty open to us adding support.

Adding support for extracting fragments associated with bits in neural fingerprints would be very interesting. I'd be interested in adding support for this, but off the top of my head, it's not clear to me how this was done. Did they have a description in the Duvenaud paper?

**Comment by tobiasharren** [comment]
Hello, thanks for your reply!Yes it would be single atoms. It could be used, when trying to interpret models, like in the lime example. One wants to determine the importance of certain fingerprint bits and then wants to map these important bits back to the fragments, which might just be one single atom. Actually it might be necessary to add more diversification than what I did in that example to really identify the exact substructures the bits refer to including hybrdidisation, thereby eliminating the smiles which are the same. But to do that we would have to go deeper into the routine and definetly swap to smiles from smarts. I will probably see if I can find out a trick, greg landrum also linked that notebook in the thread I mentioned and maybe I can come up with something. As for the neural fingerprints: They have the code in their github repository. They somehow backtrack atom activations, but there is no detailed description for that and I was not able to quickly identify the technique, I just thought you might already have something in hand for that.best regardsTobiasAm 05.03.2020 20:23 schrieb Bharath Ramsundar <notifications@github.com>:What would an environment of radius 0 correspond to? Would that just be an atom? If there's a clear use case for this, I'd be pretty open to us adding support.
Adding support for extracting fragments associated with bits in neural fingerprints would be very interesting. I'd be interested in adding support for this, but off the top of my head, it's not clear to me how this was done. Did they have a description in the Duvenaud paper?

**Comment by rbharath** [comment]
Awesome, in that case, feel free to open a PR. It would help a lot for review if you could add test cases that show how the new feature would be used and test basic functionality. Major bonus points for a short IPython notebook example of usage :)

For the neural fingerprints, we should figure out the technique they used. If we can do that, would love to see an implementation added!

**Comment by peastman** [comment]
In principle this should be simple.  You just loop over all fragments in the molecule, and identify which bit each one is mapped to.  RDKit clearly must have the code to do that, since it computes fingerprints.  The question is whether they expose it through any public interface.

**Comment by tobiasharren** [comment]
Yes in principle it's rather uncomplicated, you get for every bit one set of atoms, which activate this bit from rdkit. Then you have to create smarts patterns which fit these atoms, but take into account every detail like aromaticity, bond types and hybridisation. There might still be multiple fragments in the molecule fullfilling this pattern, but no pattern should occur multiple times, which makes it neccessary to create rather detailed smarts, which will be a little bit more complicated. Am 05.03.2020 21:11 schrieb peastman <notifications@github.com>:In principle this should be simple.  You just loop over all fragments in the molecule, and identify which bit each one is mapped to.  RDKit clearly must have the code to do that, since it computes fingerprints.  The question is whether they expose it through any public interface.

**Comment by peastman** [comment]
> you get for every bit one set of atoms, which activate this bit from rdkit.

Not necessarily just one.  The hashing function maps a potentially unlimited variety of fragments to a fixed number of bits, so many different fragments all map to each bit.  But only a limited number of them will appear in your particular data set.  So for each bit, you want to identify the (hopefully small) set of fragments in your data set that map to it.

**Comment by tobiasharren** [comment]
sorry, I was not precise, you get one fragment for each bit in the unfolded vector (or at least very unlikely collisions, since it has a size of something like 2^20) in the folded vector of lets say 1024 bits theres very likely collisions, especially with large molecules and a bigger radius. Am 05.03.2020 21:44 schrieb peastman <notifications@github.com>:
you get for every bit one set of atoms, which activate this bit from rdkit.

Not necessarily just one.  The hashing function maps a potentially unlimited variety of fragments to a fixed number of bits, so many different fragments all map to each bit.  But only a limited number of them will appear in your particular data set.  So for each bit, you want to identify the (hopefully small) set of fragments in your data set that map to it.

**Comment by tobiasharren** [comment]
Hi,

I have edited this method a little bit: https://github.com/greglandrum/rdkit_blog/blob/master/notebooks/Explaining%20Morgan%20Similarity.ipynb and inserted the code into the featurization file. It runs just fine for the couple of molecules I have tested yet, but sadly I cannot provide any tests.
I don't exactly know how to make a pull request, so I will attach the file here
[fingerprints.py.txt](https://github.com/deepchem/deepchem/files/4297448/fingerprints.py.txt)

I hope this is helpful nonetheless.

**Comment by KacperKubara** [comment]
Hi,
Regarding the neural fingerprints, I just wanted to add that it would be super useful to be able to use LIME with them as in this [tutorial on Tox21](https://deepchem.io/docs/notebooks/Explaining_Tox21.html). The problem as mentioned by @tobiasharren  is that we don't know at the moment how to map each bit of the fingerprint back to the mol/SMILES representation. 
The code repository for the paper gives some hints, especially [in this file](https://github.com/HIPS/neural-fingerprint/blob/master/examples/visualization.py#L175):
 
The `plot()` function in this file creates the activated parts of molecules as in the paper. 
https://github.com/HIPS/neural-fingerprint/blob/d26dae3a293a736261fee307b28b37e548c5f218/examples/visualization.py#L173-L237

Their code implementation is a bit obscure. I also struggle to understand what they did but hopefully, together we would be able to figure it out :) From what I currently see they did the following (I might be wrong tho, so please double check it!):

1. They take activations for each atom in the molecule (details below)
2. For each fingerprint index they find the "most-activating" atom that has a specific radius defined for the neighborhood
3. Plot the "most-activating" atom with its neighborhood
The atom_activations are computed using this function, I still haven't figure out what it actually is: 
https://github.com/HIPS/neural-fingerprint/blob/d26dae3a293a736261fee307b28b37e548c5f218/neuralfingerprint/build_convnet.py#L72-L96

**Comment by rbharath** [comment]
I took a first glance through this code and it's a little hard for me to understand too. The original neural fingerprint code was written in autograd, so the conversion to TensorFlow was quite challenging to pull off. I think the relevant code is in this blob:

```
    for fp_ix in range(params['fp_length']):
        print "FP {0} has linear regression coefficient {1}".format(fp_ix, last_layer_weights[fp_ix][0])
        combined_list = []
        for radius in all_radii:
            fp_activations = atom_activations[radius][:, fp_ix]
            combined_list += [(fp_activation, atom_ix, radius) for atom_ix, fp_activation in enumerate(fp_activations)]


        unique_list = remove_duplicates(combined_list, key_lambda=lambda x: x[0])
        combined_list = sorted(unique_list, key=lambda x: -x[0])


        for fig_ix in range(num_figs_per_fp):
            # Find the most-activating atoms for this fingerprint index, across all molecules and depths.
            activation, most_active_atom_ix, cur_radius = combined_list[fig_ix]
            most_activating_mol_ix = parent_molecule_dict[most_active_atom_ix]
            highlight_list_our_ixs = get_neighborhood_ixs(array_rep, most_active_atom_ix, cur_radius)
            highlight_list_rdkit = [array_rep['rdkit_ix'][our_ix] for our_ix in highlight_list_our_ixs]


            print "radius:", cur_radius, "atom list:", highlight_list_rdkit, "activation", activation
            draw_molecule_with_highlights(
                "figures/fp_{0}_highlight_{1}.pdf".format(fp_ix, fig_ix),
                train_smiles[most_activating_mol_ix],
                highlight_atoms=highlight_list_rdkit)
```

Where the `most_active_atom_ix` is computed.

**Comment by KacperKubara** [comment]
Ah, I think I have figured it out (more or less). I need this stuff for my undergrad dissertation so I could spend some time on it :) The key to understand what's happening here are `atom_activations` and how do they relate to the paper. Before that though, let's have a look at the following part of the code:
``` Python
   # For each fingerprint bit
    for fp_ix in range(params['fp_length']):
        print "FP {0} has linear regression coefficient {1}".format(fp_ix, last_layer_weights[fp_ix][0])
        combined_list = []
        # For each radius counted from the most-activating atom
        # all_radii is defined as params['fp_depth'] + 1 where fp_depth is a layer depth
        for radius in all_radii:
            # Take atom activations for the specifc radius and fingerpint bit
            # Why do we take radius like it? -> Because the layer depth is related to the neighbourhood radius (look at the paper)
            # if the layer is deeper that means that we take information from a neighbourhood with a bigger radius
            fp_activations = atom_activations[radius][:, fp_ix] 
            # Create a list of activations for each radius
            combined_list += [(fp_activation, atom_ix, radius) for atom_ix, fp_activation in enumerate(fp_activations)]

        unique_list = remove_duplicates(combined_list, key_lambda=lambda x: x[0])
        # Sort by decreasing activation
        combined_list = sorted(unique_list, key=lambda x: -x[0])

        for fig_ix in range(num_figs_per_fp):
            # Find the most-activating atoms for specific fingerprint bit, across all molecules and depths (radius).
            activation, most_active_atom_ix, cur_radius = combined_list[fig_ix]
            # Take a molecule to which the most_active_atom_ix ID belongs
            most_activating_mol_ix = parent_molecule_dict[most_active_atom_ix]
            # Get the neighbouring atom within a certain radius to most_active_atom_ix 
            highlight_list_our_ixs = get_neighborhood_ixs(array_rep, most_active_atom_ix, cur_radius)
            # Just some conversion for RDKit
            highlight_list_rdkit = [array_rep['rdkit_ix'][our_ix] for our_ix in highlight_list_our_ixs]

            print "radius:", cur_radius, "atom list:", highlight_list_rdkit, "activation", activation
            # fp_ix is a fingerprint bit
            # fig_ix controls how many molecule substructures we want to find for fp_ix
            # fig_ix=0 has the biggest activation value, fig_ix=num_figs_per_fp the smallest activation value
            draw_molecule_with_highlights(
                "figures/fp_{0}_highlight_{1}.jpg".format(fp_ix, fig_ix),
                train_smiles[most_activating_mol_ix],
                highlight_atoms=highlight_list_rdkit)
```
#### General understanding of the code above
I've added some comments here to understand what's going on. Basically what I think is happening here:

1. For each fingerprint bit:

- We take `atom_activations` for the specifc _fp_ bit and for varying radius which defines the neighbourhood of the most activated atom. Radius is basically a layer depth of the network which is also explained in the paper: 

> The size of the substructures represented by each index depends on the depth of the network. Thus the number of layers is referred to as the ‘radius’ of the fingerprints.

In other words, we take activations for specific _fp_ bit across all network layers.

- Then is just a matter of transforming the data to plot it (more details in the comments). We basically take top _k_ molecule substructures with the highest activations for specific fingerprint bit.

### Now, what the 🦆 is the `atom_activations` ?!
Code below is responsible for generating those enigmatic `atom_activations`:
``` Python
    def output_layer_fun_and_atom_activations(weights, smiles):
        """Computes layer-wise convolution, and returns a fixed-size output."""
        # This code will compute these activations for all molecules that were passed at once
        # Handy function to generate all  molecular features for us
        array_rep = array_rep_from_smiles(tuple(smiles))
        atom_features = array_rep['atom_features']
        bond_features = array_rep['bond_features']

        all_layer_fps = []
        atom_activations = []
        def write_to_fingerprint(atom_features, layer):
            # Output weights for specific layer
            cur_out_weights = parser.get(weights, ('layer output weights', layer))
            # Output bias for specific layer
            cur_out_bias    = parser.get(weights, ('layer output bias', layer))
            # Apply softmax after a linear layer - im not certain what are the dimensions
            atom_outputs = softmax(cur_out_bias + np.dot(atom_features, cur_out_weights), axis=1)
            # We get activations for specifc layer for all molecules
            # I think we got activations that have fp size for a certain radius (layer depth)
            atom_activations.append(atom_outputs)
            # We are not interested in this stuff below -> we dont use that in the code for plotting
            # Sum over all atoms within a moleclue:
            layer_output = sum_and_stack(atom_outputs, array_rep['atom_list'])
            all_layer_fps.append(layer_output)

        num_layers = len(num_hidden_features)
        # For layer depth (radius of neighbourhood)
        for layer in xrange(num_layers):
            # Get the atom_activations for specific radius for all molecules
            write_to_fingerprint(atom_features, layer)
            # Propagate the results to another layer (???)
            atom_features = update_layer(weights, layer, atom_features, bond_features, array_rep,
                                         normalize=normalize)
        write_to_fingerprint(atom_features, num_layers)
        return sum(all_layer_fps), atom_activations, array_rep
```
#### General understanding of the code above

1. Get atom features for molecules that we passed as an argument
1. For each layer (neighbourhood radius):
- Get bias and weight for layer
- Apply linear layer and then softmax function to get atom_outputs (I believe that size should be (fp_size, molecule_count) but I might be wrong)
- Append the output to `atom_activations`
- Use `update_layer()` to pass the values to another layer
3. Return the atom_activations

So hopefully my small essay will make the code more clear. The code is quite difficult to understand so please bear in mind that is possible that I've made a mistake somewhere.
